1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency with the PowerShell ForEach-Object cmdlet, without having to obtain and load a separate module, or deal with PowerShell jobs unless I want to.
+
+## Specification
+
+There will be two new parameter sets added to the existing ForeEach-Object cmdlet to support both synchronous and asynchronous operations for parallel script block execution.
+For the synchronous case, the `ForEach-Object` cmdlet will not return until all parallel executions complete.
+For the asynchronous case, the `ForEach-Object` cmdlet will immediately return a PowerShell job object that contains child jobs of each parallel execution.
+
+### Implementation details
+
+Implementation will be similar to the ThreadJob module.
+Script block execution will be run for each piped input on a separate thread and runspace.
+The number of threads that run at a time will be limited by a `-ThrottleLimit` parameter with a default value.
+Piped input that exceeds the allowed number of threads will be queued until a thread is available.
+For synchronous operation, a `-Timeout` parameter will be available that terminates the wait for completion after a specified time.
+Without a `-Timeout` parameter, the cmdlet will wait indefinitely for completion.
+
+### Synchronous parameter set

```

Comment:
I think it would be helpful to specify the following behavior for this parameter set:

1. How is an exception from one but not all scriptblock invocations handled?  Are the other non-throwing scriptblocks allowed to complete?  Is the exception ever thrown?  Is it converted to an error instead?
2. How are exceptions from multiple scriptblock invokations handled?  Are they combined into an AggregateException?
3. How are Verbose/Warning/Error streams from scriptblock invocations handled?  Are they multiplexed and passed through so they appear to the user?  If so, how can the user correlate those streams by invocation?  Some sort of ID prefix?
4. Does ForEach-Object write its progress to the Progress stream?
5. How are Progress streams from scriptblock invocations handled?  Do they become child activities to a ForEach-Object parent activity?

Reply:
I think in current design it will work "as is" - based on runspace affinity. I expect we will get sometimes "interesting" side effects.

Reply:
The runspace affinity can be completely avoid in the implementation, by passing the AST instead of the ScriptBlock to another Runspace, then get the script block by `Ast.GetScriptBlock()`

Reply:
Those are good questions that should be answered in this RFC.

Reply:
I agree.  I'll add answers to the RFC for these questions:.

Reply:
>The runspace affinity can be completely avoid in the implementation, by passing the AST instead of the ScriptBlock to another Runspace, then get the script block by Ast.GetScriptBlock()

I don't see how this is true, in general.  For example, in

```PowerShell
$g = Get-SomeScriptBlockGenerator

1..10 | ForEach-Object -Parallel { . $using:g.GetNextScriptBlock() }
```

would the the scriptblock returned by .GetNextScriptBlock() be silently converted to its .Ast counterpart?

Reply:
The parser will not allow the example shown above, the $using directive only supports variable expressions.  And generating a new script block from the Ast will only be done for ScriptBlock using variable types.

But it is still possible to pass in a script block variable aligned to a different runspace.  e.g.,
```powershell
ForEach-Parallel -Parallel { $g = $using:g; $g.GetNextScriptBlock() }
```
We won't be trying to prevent all possible cases of passing a variable reference to a script block and using it unsafely.  But we can document safe and unsafe ways to use passed in variable references.